godot-core/src/builtin/basis.rs

```diff
@@ -0,0 +1,980 @@
+    #[must_use]
+    pub const fn from_diagonal(x: f32, y: f32, z: f32) -> Self {
+        Self {
+            rows: [
+                Vector3::new(x, 0.0, 0.0),
+                Vector3::new(0.0, y, 0.0),
+                Vector3::new(0.0, 0.0, z),
+            ],
+        }
+    }
+
+    /// Create a diagonal matrix from the given values.
+    ///
+    /// _Godot equivalent: `Basis.from_scale(Vector3 scale)`
+    #[must_use]
+    pub const fn from_scale(scale: Vector3) -> Self {
+        Self::from_diagonal(scale.x, scale.y, scale.z)
+    }
+
+    /// Create a new basis from 3 row vectors. These are *not* basis vectors.
+    #[must_use]
+    pub const fn from_rows(x: Vector3, y: Vector3, z: Vector3) -> Self {
+        Self { rows: [x, y, z] }
+    }
+
+    /// Create a new basis from 3 column vectors.
+    #[must_use]
+    pub const fn from_cols(x: Vector3, y: Vector3, z: Vector3) -> Self {
+        Self::from_rows_array(&[x.x, y.x, z.x, x.y, y.y, z.y, x.z, y.z, z.z])
+    }
+
+    #[must_use]
+    const fn from_rows_array(rows: &[f32; 9]) -> Self {
+        let [x0, y0, z0, x1, y1, z1, x2, y2, z2] = rows;
+        Self {
+            rows: [
+                Vector3::new(*x0, *y0, *z0),
+                Vector3::new(*x1, *y1, *z1),
+                Vector3::new(*x2, *y2, *z2),
+            ],
+        }
+    }
+
+    /// Create a `Basis` from an axis and angle.
+    ///
+    /// _Godot equivalent: `Basis(Vector3 axis, float angle)`_
+    #[must_use]
+    pub fn from_axis_angle(axis: Vector3, angle: f32) -> Self {
+        Mat3::from_axis_angle(axis.to_glam(), angle).to_front()
+    }
+
+    /// Create a `Basis` from a `Quaternion`.
+    ///
+    /// _Godot equivalent: `Basis(Quaternion from)`_
+    #[must_use]
+    pub fn from_quat(quat: Quaternion) -> Self {
+        Mat3::from_quat(quat.to_glam()).to_front()
+    }
+
+    /// Create a `Basis` from three angles `a`, `b`, and `c` interpreted
+    /// as Euler angles according to the given `EulerOrder`.
+    ///
+    /// _Godot equivalent: `Basis.from_euler(Vector3 euler, int order)`_
+    #[must_use]
+    pub fn from_euler(order: EulerOrder, a: f32, b: f32, c: f32) -> Self {
+        // Translated from "Basis::from_euler" in
+        // https://github.com/godotengine/godot/blob/master/core/math/basis.cpp
+
+        // We can't use glam to do these conversions since glam uses intrinsic rotations
+        // whereas godot uses extrinsic rotations.
+        // see https://github.com/bitshifter/glam-rs/issues/337
+        let xmat =
+            Basis::from_rows_array(&[1.0, 0.0, 0.0, 0.0, a.cos(), -a.sin(), 0.0, a.sin(), a.cos()]);
+        let ymat =
+            Basis::from_rows_array(&[b.cos(), 0.0, b.sin(), 0.0, 1.0, 0.0, -b.sin(), 0.0, b.cos()]);
+        let zmat =
+            Basis::from_rows_array(&[c.cos(), -c.sin(), 0.0, c.sin(), c.cos(), 0.0, 0.0, 0.0, 1.0]);
+
+        match order {
+            EulerOrder::XYZ => xmat * (ymat * zmat),
```

Comment:
Matrix multiplication is associative, so the parentheses are not needed here. It's otherwise confusing why only one variant would use right-associative multiplication.

Reply:
you're right, this comes from the godot code. not entirely sure why they did it that way

Reply:
Maybe related to numerical precision? Although that would still not explain why exactly this one.
I would try without parentheses, if we run into issues we can still change it.